feat(sort): allowing sorting by multiple columns

relair opened the pull request:
Extended the Sort interface for active to allow string array and SortDirection to allow dictionary object. Adding matMultiColumn property to MatSort that will allow multi column sorting. Updating MatTableDataSource to handle multi column sorting. Adding a counter for sort-header that shows sorting precedence.

Breaking change: Because of extended interface of MatSort, properties active and direction now need to be cast to string for existing usages. Fixes #7226.

GO3LIN commented:
@jelbourn @andrewseguin Have you got the time to review this PR, I saw the the issue is labeled P3, hence it's an important PR. I think that the community in a real need of this feature.

joshorvis commented:
> @jelbourn @andrewseguin Have you got the time to review this PR, I saw the the issue is labeled P3, hence it's an important PR. I think that the community is in a real need of this feature.

<--- Member of the community in real need of this feature...   Please with cherries on top?

jelbourn commented:
@relair we appreciate time time spent working on this. I took an initial look and have some high-level comments:

* Rather than adding multi-column sorting to the existing `MatSort` directive, we'd prefer to introduce a new `MatMultiSort` directive. The new directive would share relevant code with `MatSort`, but would let people only pull in the specific behavior they want in their app.
* I'm not sure the number indicator next to the arrow indicators is a good fit for the component. I've reached out to some UX designers inside Google to see what they would recommend here. In the one app I've looked at that already supports multi-column sorting (GCP), they don't use a visual indicator at all for the non-primary columns.
* We have to make sure that screen-reader users are able to interact with the sorting effectively. As it stands now, the extra number in the header doesn't do enough to communicate the sort state. Unfortunately the W3C doesn't yet have any guidance on multi-column sorting, so there's no obvious solution. My best guess would be adding an `aria-describedby` to the sort button that describes the sort status (this would have to come from `MatSortHeaderIntl`). I'd recommend trying the component out in a screen-reader (either NVDA on Windows or VoiceOver on macOS) and seeing how it goes. If you're unfamiliar with using screen-readers, [Rob Dodson has some great videos in his a11ycasts series](https://www.youtube.com/playlist?list=PLNYkxOF6rcICWx0C9LVWWVqvHlYJyqw7g).

relair commented:
For now I will make MatSortHeader work for both MatSort and MatMultiSort as I think it would be awkward for the user to need to define different directive name for sort headers when using MatMultiSort although I could change it too.
For now I will leave those if statements in _isSorted and _updateArrowDirection, although I could change it to something implementing strategy pattern (where strategy would be picked in the constructor based on which MatSort is available).

relair commented:
I have finished the refactoring, I did create MatMultiSort that shares some code with MatSort  trough base class. I also refactored MatSortHeader so it is using strategy pattern to chose the logic which needs to use (no more if statements everywhere), I have also fixed aria description for MatSortHeader so whenever it is sorted by several columns multiple headers would just have description that says it is sorted ascending or descending.

relair commented on src/material-examples/table-http/table-http-example.ts:
```diff
@@ -38,7 +38,7 @@ export class TableHttpExample implements OnInit {
           this.isLoadingResults = true;
           return this.exampleDatabase!.getRepoIssues(
-            this.sort.active, this.sort.direction, this.paginator.pageIndex);
+            this.sort.active as string, this.sort.direction as string, this.paginator.pageIndex);
```
This shows the breaking change - as this.sort.active can be something else than just string it required casting to string. I don't really see any way around it other than having different properties for multi sort, but it would introduce more chaos.

carlosvpp commented:
Hi! Is there any plans on merging this branch any time soon? I've testing the implementation and it's been working great for me. Good job and thanks @relair

the-avid-engineer commented:
Pleeeeeeeaaase!

hughanderson4 commented:
ALL THIS PULL REQ NEEDS IS ONE LABEL AND WE CAN HAVE MULTI COLUMN SORTING!

WE CAN DO THIS!!!!

Only1MrAnderson commented:
Been watching this for a while it will be sad to see if this PR becomes outdated and kicked to the side

hughanderson4 commented:
> Hi! Is there any plans on merging this branch any time soon? I've testing the implementation and it's been working great for me. Good job and thanks @relair

@carlosvpp can you please explain how you tested this?  I can't seem to find a way to merge this into my Material fork.

andrewseguin commented:
Hey @relair, thanks for submitting this pull request. I apologize that it has taken so long for us to dive into it and get it merged into the library. You did a great job with this and I think we can work with this to get the functionality in.

I looked over the code and I recognized that we have an existing issue, which is likely what you found as well. The `MatSortable` determines if it is active by retrieving information from the container and then performing its own logic. You resolved this by creating strategies which I think was a valid solution, but I wonder if we can go another step forward and contain this logic within the `MatSort` and `MatMultiSort` directly.

I'm thinking we can have the `MatSortable` request its state from the container, which will tell it whether it is active, its direction, etc. That way, the `MatSortable` only reflects the state it is given rather than trying to determine it itself.

Another bit of contention the UX that I think we can solve if we allow the user to provide a custom template to be used to the right of the arrow.

I'm going to try and come up with a PR that inverts that state management a bit and we can go from there. Once again I'm sorry that we drop the ball on reviewing this sooner. We got caught up in helping Angular with Ivy but I want to make sure we give some love to people wanting to help us and contribute to make Material awesome.

andrewseguin commented:
I'm going to split this pull request up into three stages. 

1. Refactor the sort header so that its sort container determines and provides its state (e.g. it is up to the sort container to determine if the sort header is sorted, not the sort header determining this based on sort container properties).
2. Add a new feature that allows the user to provide their own `ng-template` to proceed the sort arrow, which can be used for things like a multi-sort level/count indicator.
3. Add a new `MatMultiSort` that is a new sort container.

relair commented:
Hey @andrewseguin, thank you for looking into getting this feature done. I did notice that I did that pull request at a bit unfortunate time as the new angular version was just being released, so I was waiting patiently until someone will find some time.
For my work I was cautious about doing big changes, so for example I didn't move the logic out of sort-header, even after wrapping it in strategies, so it will at least be still in the same file. I agree however that just getting the strategy directly from MatSort or MatMultiSort or some generic interface for them would be better than those if statements for selecting strategy.
Hopefully you will be able to change the interfaces so it will be easier to swap the sorting implementation. Probably you will end up with more breaking changes because of it, but it will probably be worth it in the end. I guess the feature will need to go to the next big version, because of breaking changes?

andrewseguin commented:
Fortunately it looks like we'll avoid any breaking changes (except to the constructor of `MatSort` which will only affect users who extend it). However, we are already in the mode of targeting the next major so I expect this will land there. 

Thanks again for starting us off on this feature, we are now talking about moving a lot of the core logic of the sort manager to the CDK and make it easier for users to extend and swap things around.

Feel free to take a look at the first step on this initial PR: https://github.com/angular/material2/pull/15171

Next steps after that will be to allow a custom template to be provided to follow the sort arrow (for the multi sort counter), and then a pull request to include your multi-sort using the new refactored setup

raghufsmk commented:
Hi.. Could you please tell me when this feature will be merged to master?

Thanks in advance

andrewseguin commented:
@raghufsmk I probably will not have the time to personally message you that this is ready, but you are welcome to keep an eye on issue to see when it is closed, which means it is merged in master

Noppey commented:
How does it take months to get a review done? We are also waiting for this feature.

relair commented:
Hmm, why is CLA complaining again?

relair commented:
@andrewseguin Hi Andew, what is the progress on pulling this request? Do I need to do something more with that pull request that could help?

andrewseguin commented:
@relair Hey sorry for the delay, we've had to shift a lot of our work into Ivy-related issues. 

Right now we're waiting on getting [this PR in](https://github.com/angular/material2/pull/15171), which is based off your work and lets us decouple the `MatSort` and `MatSortHeader`

Noppey commented:
There is no activity in the PR that is apparently a blocker for this one? Except some emoticon labelling...
If it's gonna be 2021 before this feature is released, that's obviously up to you and totally fine, but please let us know so we can take a different path for these kind of functionalities.

andrewseguin commented:
@Noppey This is at a lower priority than getting Ivy out the door, but we still very much value this use case. We're currently considering a bigger change where we bring some of this logic into the CDK, which add a bit more of a delay since we need to make sure we design it right.

hughanderson4 commented:
there's no good reason for this to be stalled out.
been waiting nearly a year.
we were told to wait until next major release.  that has come and gone.
please it is terrible community experience to wait this long for good functionality to be merged.

BRM013 commented:
Can we get some examples on how to use the 'matSortActive' and 'matSortDirection' inputs on the MatMultiSort directive please. I'm trying to define a default sort in a template and I can't seem to define the structure correctly and get errors from .isSortDirectionValid().

himanshugarg574 commented:
Hey guys, Great work with IVY. Now that it is released, can we get an ETA on this ? Thanks.

andrewseguin commented:
This PR helped to kick off design work towards this feature. I'll close it now in favor of those new PRs to come. It's on our backlog but not a priority unfortunately. Thanks again for the work on this, it really helped us identify how we can decouple the sort and header so that they can be reusable. We really value the work and I'm sorry that we could not put more resources into it right now
